File: forms/who-emergency-first-aid-form/back-end-with-loco/tests/engine/cfar_validator_test.rs
use who_emergency_first_aid_form_loco_crate::engine::cfar_rules::cfar_rules;
use who_emergency_first_aid_form_loco_crate::engine::cfar_validator::validate_cfar;
use who_emergency_first_aid_form_loco_crate::engine::types::*;

fn empty() -> AssessmentData {
    AssessmentData::default()
}

/// Build a fully-completed, minimally valid encounter record.
fn fully_completed_encounter() -> AssessmentData {
    let mut d = AssessmentData::default();
    d.patient_identification.patient_name = "DOE, Jane".to_string();
    d.patient_identification.sex = "female".to_string();
    d.referral_transport.referral_facility.name = "St James's Hospital ED".to_string();
    d.referral_transport.event_date_time = "2026-06-01T08:00".to_string();
    d.referral_transport.departure_date_time = "2026-06-01T08:15".to_string();
    d.situation.medical = true;
    d.situation.pregnant = "no".to_string();
    d.situation.what_happened = "Patient slipped on a wet floor.".to_string();
    d.background.past_medical_and_surgical_history = "None.".to_string();
    d.background.current_medications_or_allergies = "None.".to_string();

    d.major_bleeding.assessment_normal = true;
    d.major_bleeding.interventions.none = true;
    d.airway.assessment_normal = true;
    d.airway.interventions.none = true;
    d.breathing.assessment_normal = true;
    d.breathing.interventions.none = true;
    d.circulation.assessment_normal = true;
    d.circulation.interventions.none = true;
    d.disability.assessment_normal = true;
    d.disability.interventions.none = true;
    d.exposure.assessment_normal = true;
    d.exposure.interventions.none = true;
    d.exposure.medication_taken_none = true;

    d.recommendations.transport_plan = "Transport by ambulance.".to_string();
    d.responder_details.name = "John Smith".to_string();
    d.responder_details.signature = "JS".to_string();
    d.responder_details.contact_information = "+44 7700 900123".to_string();
    d.responder_details.cfar_organization = "Red Cross".to_string();
    d
}

#[test]
fn empty_form_is_incomplete_and_skips_conditional_rules() {
    let data = empty();
    let r = validate_cfar(&data);
    assert!(!r.complete);
    assert!(r.total_required > 0);
    assert_eq!(r.total_satisfied, 0);
    assert_eq!(r.missing.len() as u32, r.total_required);
    assert!(r.missing.iter().any(|m| m.id == "PI-01"));
    // Conditional rules must NOT apply on an empty form.
    assert!(!r.missing.iter().any(|m| m.id == "MB-03"));
    assert!(!r.missing.iter().any(|m| m.id == "RC-02"));
}

#[test]
fn assessment_normal_or_finding_text_satisfies_assessment_rule() {
    let mut data = empty();
    data.major_bleeding.assessment_normal = true;
    data.airway.assessment_normal = true;
    let r = validate_cfar(&data);
    assert!(!r.missing.iter().any(|m| m.id == "MB-01"));
    assert!(!r.missing.iter().any(|m| m.id == "AW-01"));

    let mut data2 = empty();
    data2.major_bleeding.assessment_findings = "Bright red bleeding, left thigh".into();
    data2.airway.assessment_findings = "Audible stridor".into();
    let r2 = validate_cfar(&data2);
    assert!(!r2.missing.iter().any(|m| m.id == "MB-01"));
    assert!(!r2.missing.iter().any(|m| m.id == "AW-01"));

    let mut data3 = empty();
    data3.major_bleeding.interventions.none = true;
    let r3 = validate_cfar(&data3);
    assert!(!r3.missing.iter().any(|m| m.id == "MB-02"));
}

#[test]
fn tourniquet_without_time_fires_mb03() {
    let mut data = empty();
    data.major_bleeding.assessment_findings = "Massive bleeding, right arm".into();
    data.major_bleeding.interventions.tourniquet = true;
    let r = validate_cfar(&data);
    assert!(r.missing.iter().any(|m| m.id == "MB-03"));
}

#[test]
fn other_precaution_without_details_fires_rc02() {
    let mut data = empty();
    data.recommendations.precautions.other = true;
    let r = validate_cfar(&data);
    assert!(r.missing.iter().any(|m| m.id == "RC-02"));
}

#[test]
fn fully_completed_encounter_is_complete() {
    let data = fully_completed_encounter();
    let r = validate_cfar(&data);
    assert!(r.complete, "missing: {:?}", r.missing);
    assert_eq!(r.total_satisfied, r.total_required);
}

#[test]
fn rule_ids_are_unique() {
    let mut ids: Vec<&str> = cfar_rules().iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

File: forms/who-emergency-first-aid-form/front-end-with-svelte/src/lib/engine/cfar-validator.test.ts
import { describe, expect, it } from 'vitest';
import { cfarRules } from './cfar-rules';
import { validateCfar } from './cfar-validator';
import { detectFlaggedIssues } from './flagged-issues';
import type { AssessmentData } from './types';

function emptyAssessment(): AssessmentData {
	return {
		patientIdentification: {
			patientName: '',
			dateOfBirth: '',
			age: null,
			sex: '',
			patientContactInformation: '',
			contactPerson: { name: '', contactInformation: '' }
		},
		referralTransport: {
			referralFacility: { name: '', focalPoint: '', phoneNumber: '' },
			ambulance: { name: '', focalPoint: '', phoneNumber: '' },
			eventDateTime: '',
			departureDateTime: ''
		},
		situation: {
			medical: false,
			trauma: false,
			pregnant: '',
			whatHappened: ''
		},
		background: {
			pastMedicalAndSurgicalHistory: '',
			currentMedicationsOrAllergies: ''
		},
		majorBleeding: {
			assessmentNormal: false,
			assessmentFindings: '',
			interventions: {
				directPressure: false,
				deepWoundPacking: false,
				tourniquet: false,
				tourniquetApplicationTime: '',
				uterineMassage: false,
				none: false
			}
		},
		airway: {
			assessmentNormal: false,
			assessmentFindings: '',
			interventions: {
				neckImmobilization: false,
				headTiltChinLift: false,
				jawThrust: false,
				chokingCare: false,
				none: false
			}
		},
		breathing: {
			assessmentNormal: false,
			assessmentFindings: '',
			interventions: {
				maintainedPositionOfComfort: false,
				none: false
			}
		},
		circulation: {
			assessmentNormal: false,
			assessmentFindings: '',
			interventions: {
				pelvicBinder: false,
				controlMinorBleeding: false,
				fractureCare: false,
				oralHydration: false,
				leftLateralPosition: false,
				none: false
			}
		},
		disability: {
			assessmentNormal: false,
			assessmentFindings: '',
			interventions: {
				spinalImmobilisation: false,
				glucoseGiven: false,
				seizureCare: false,
				highTemperatureCare: false,
				lowTemperatureCare: false,
				none: false
			}
		},
		exposure: {
			assessmentNormal: false,
			assessmentFindings: '',
			interventions: {
				recoveryPosition: false,
				burnCare: false,
				woundCare: false,
				drowningCare: false,
				snakebiteCare: false,
				none: false
			},
			medicationTakenNone: false,
			medicationTakenDetails: ''
		},
		recommendations: {
			transportPlan: '',
			problemsAnticipated: '',
			otherConcerns: '',
			precautions: {
				highlyInfectiousDisease: false,
				spinalImmobilization: false,
				possibleFracture: false,
				fallRisk: false,
				alteredMentalStatus: false,
				other: false,
				otherDetails: ''
			}
		},
		responderDetails: {
			name: '',
			signature: '',
			contactInformation: '',
			cfarOrganization: ''
		}
	};
}

/**
 * Build a minimal-but-complete encounter for a low-acuity patient: all
 * CABCDE assessments marked Normal, all interventions ticked None, no
 * precautions, mandatory text fields filled with the simplest content.
 */
function minimalCompleteAssessment(): AssessmentData {
	const d = emptyAssessment();

	d.patientIdentification.patientName = 'DOE, Jane';
	d.patientIdentification.sex = 'female';

	d.referralTransport.referralFacility.name = 'District Hospital';
	d.referralTransport.eventDateTime = '2026-04-30T10:00';
	d.referralTransport.departureDateTime = '2026-04-30T10:15';

	d.situation.medical = true;
	d.situation.pregnant = 'no';
	d.situation.whatHappened = 'Patient reported chest discomfort at home around 09:50.';

	d.background.pastMedicalAndSurgicalHistory = 'None';
	d.background.currentMedicationsOrAllergies = 'None';

	d.majorBleeding.assessmentNormal = true;
	d.majorBleeding.interventions.none = true;

	d.airway.assessmentNormal = true;
	d.airway.interventions.none = true;

	d.breathing.assessmentNormal = true;
	d.breathing.interventions.none = true;

	d.circulation.assessmentNormal = true;
	d.circulation.interventions.none = true;

	d.disability.assessmentNormal = true;
	d.disability.interventions.none = true;

	d.exposure.assessmentNormal = true;
	d.exposure.interventions.none = true;
	d.exposure.medicationTakenNone = true;

	d.recommendations.transportPlan = 'Transport to District Hospital ED by ambulance.';

	d.responderDetails.name = 'Alex Responder';
	d.responderDetails.signature = 'A. Responder';
	d.responderDetails.contactInformation = '+44 7700 900000';
	d.responderDetails.cfarOrganization = 'Community First Aid Network';

	return d;
}

describe('CFAR Rule Set', () => {
	it('has a non-empty rule set', () => {
		expect(cfarRules.length).toBeGreaterThan(0);
	});

	it('all rule IDs are unique', () => {
		const ids = cfarRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});

	it('every rule references a known section', () => {
		const validSections = new Set([
			'patientIdentification',
			'referralTransport',
			'situation',
			'background',
			'majorBleeding',
			'airway',
			'breathing',
			'circulation',
			'disability',
			'exposure',
			'recommendations',
			'responderDetails'
		]);
		for (const r of cfarRules) {
			expect(validSections.has(r.section)).toBe(true);
		}
	});
});

describe('CFAR Validator', () => {
	it('marks an empty assessment as incomplete', () => {
		const result = validateCfar(emptyAssessment());
		expect(result.complete).toBe(false);
		expect(result.totalSatisfied).toBe(0);
		expect(result.missing.length).toBe(result.totalRequired);
	});

	it('marks a minimal but complete encounter as complete', () => {
		const result = validateCfar(minimalCompleteAssessment());
		expect(result.complete).toBe(true);
		expect(result.missing).toHaveLength(0);
		expect(result.totalSatisfied).toBe(result.totalRequired);
	});

	it('CABCDE assessment can be answered by either Normal or findings', () => {
		const d = minimalCompleteAssessment();
		d.airway.assessmentNormal = false;
		d.airway.assessmentFindings = 'Slight stridor on inspiration.';
		const result = validateCfar(d);
		const ids = result.missing.map((m) => m.id);
		expect(ids).not.toContain('AW-01');
	});

	it('intervention block requires at least one selection or "None"', () => {
		const d = minimalCompleteAssessment();
		d.circulation.interventions.none = false;
		const result = validateCfar(d);
		const ids = result.missing.map((m) => m.id);
		expect(ids).toContain('CR-02');
	});

	it('tourniquet ticked: requires application time', () => {
		const d = minimalCompleteAssessment();
		d.majorBleeding.assessmentNormal = false;
		d.majorBleeding.assessmentFindings = 'Arterial bleeding to right thigh.';
		d.majorBleeding.interventions.none = false;
		d.majorBleeding.interventions.directPressure = true;
		d.majorBleeding.interventions.tourniquet = true;
		// time intentionally left blank
		const result = validateCfar(d);
		const ids = result.missing.map((m) => m.id);
		expect(ids).toContain('MB-03');
	});

	it('tourniquet not ticked: time field is not required', () => {
		const d = minimalCompleteAssessment();
		const result = validateCfar(d);
		const ids = result.missing.map((m) => m.id);
		expect(ids).not.toContain('MB-03');
	});

	it('"Other" precaution: requires details', () => {
		const d = minimalCompleteAssessment();
		d.recommendations.precautions.other = true;
		const result = validateCfar(d);
		const ids = result.missing.map((m) => m.id);
		expect(ids).toContain('RC-02');
	});

	it('"Other" precaution not ticked: details field not required', () => {
		const d = minimalCompleteAssessment();
		const result = validateCfar(d);
		const ids = result.missing.map((m) => m.id);
		expect(ids).not.toContain('RC-02');
	});

	it('exposure medication: either "None" or details satisfies the rule', () => {
		const d = minimalCompleteAssessment();
		d.exposure.medicationTakenNone = false;
		d.exposure.medicationTakenDetails = 'Paracetamol 500mg taken at 09:30';
		const result = validateCfar(d);
		const ids = result.missing.map((m) => m.id);
		expect(ids).not.toContain('EX-03');
	});

	it('groups missing items by section', () => {
		const result = validateCfar(emptyAssessment());
		const sectionKeys = new Set(result.sections.map((s) => s.section));
		expect(sectionKeys.has('patientIdentification')).toBe(true);
		expect(sectionKeys.has('responderDetails')).toBe(true);
		// Each section's "missing" array length must equal required - satisfied
		for (const s of result.sections) {
			expect(s.missing.length).toBe(s.required - s.satisfied);
		}
	});
});

describe('Flagged Issues', () => {
	it('returns no flags for a benign minimal-complete encounter', () => {
		const flags = detectFlaggedIssues(minimalCompleteAssessment());
		expect(flags).toHaveLength(0);
	});

	it('flags major bleeding without intervention as urgent', () => {
		const d = minimalCompleteAssessment();
		d.majorBleeding.assessmentNormal = false;
		d.majorBleeding.assessmentFindings = 'Heavy bleeding from leg laceration.';
		d.majorBleeding.interventions.none = false;
		const flags = detectFlaggedIssues(d);
		const fired = flags.find((f) => f.id === 'FLAG-MB-001');
		expect(fired).toBeDefined();
		expect(fired?.priority).toBe('urgent');
		expect(flags[0].priority).toBe('urgent');
	});

	it('flags tourniquet without time as urgent', () => {
		const d = minimalCompleteAssessment();
		d.majorBleeding.assessmentNormal = false;
		d.majorBleeding.assessmentFindings = 'Arterial bleed.';
		d.majorBleeding.interventions.none = false;
		d.majorBleeding.interventions.directPressure = true;
		d.majorBleeding.interventions.tourniquet = true;
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-MB-002' && f.priority === 'urgent')).toBe(true);
	});

	it('flags abnormal airway without intervention as urgent', () => {
		const d = minimalCompleteAssessment();
		d.airway.assessmentNormal = false;
		d.airway.assessmentFindings = 'Patient gurgling.';
		d.airway.interventions.none = false;
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-AW-001' && f.priority === 'urgent')).toBe(true);
	});

	it('flags pregnancy as high', () => {
		const d = minimalCompleteAssessment();
		d.situation.pregnant = 'yes';
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-PR-001' && f.priority === 'high')).toBe(true);
	});

	it('flags highly infectious disease precaution as high', () => {
		const d = minimalCompleteAssessment();
		d.recommendations.precautions.highlyInfectiousDisease = true;
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-PC-001' && f.priority === 'high')).toBe(true);
	});

	it('flags possible fracture as medium', () => {
		const d = minimalCompleteAssessment();
		d.recommendations.precautions.possibleFracture = true;
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-PC-003' && f.priority === 'medium')).toBe(true);
	});

	it('flags glucose given as low', () => {
		const d = minimalCompleteAssessment();
		d.disability.interventions.none = false;
		d.disability.interventions.glucoseGiven = true;
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-DS-002' && f.priority === 'low')).toBe(true);
	});

	it('sorts flags by priority (urgent > high > medium > low)', () => {
		const d = minimalCompleteAssessment();
		d.majorBleeding.assessmentNormal = false;
		d.majorBleeding.assessmentFindings = 'Heavy bleeding.';
		d.majorBleeding.interventions.none = false;
		d.situation.pregnant = 'yes';
		d.recommendations.precautions.possibleFracture = true;
		d.disability.interventions.none = false;
		d.disability.interventions.glucoseGiven = true;
		const flags = detectFlaggedIssues(d);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 } as const;
		const priorities = flags.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});
